[TPHD] Add bloom graphic pack

KoB-Kirito opened the pull request:
Inspired by the work of @skalfate - https://github.com/cemu-project/cemu_graphic_packs/issues/508#issuecomment-939681608

I tinkered a lot with the values, tested them in different areas of the game and condensed them in useful presets.
Changing the bloom improves this game so much, it should be available to everyone!

`Changes the bloom in the game, which is used aggressively in the shadow world, where it has a strong yellow cast.
Can also remove the yellow tint by shifting the bloom color to blue. A too high value can cause a blue cast outside of the shadow world. Up to 50% is safe.`

Here a comparison.
Native bloom:
![bloom native](https://github.com/user-attachments/assets/9d10530a-7ee7-4daf-8a5d-fd728056683d)

50% bloom, 50% blue shift:
![bloom 50 blue shift 50](https://github.com/user-attachments/assets/6543ab9e-7fdb-498e-8862-b4b55675643a)

Bloom disabled:
![bloom disabled](https://github.com/user-attachments/assets/60626a37-723f-47b3-a8db-97f9dfcf87c7)

KoB-Kirito commented:
I found more shaders connected to bloom, dived deep into glsl to understand them and improved this pack a lot!

- I apply the bloom level to the mask directly now, which results in a more even distribution of bloom when lowered. That also enables levels above 100% again.
- I found out how the color tint is applied (which is a static color that differs from area to area btw) and interpolate from that to an even luminance. Basically you can choose freely between intended tint and no tint at all safely without blue tint elsewhere now. (Let me know if you need custom color shift back, it didn't feel very helpful though)
- I found the shaders that color grade the shadow world darker and desaturate it. Not directly bound to bloom, but fits in this pack I think. For most that just find the shadow world too depressing, changing these settings is more useful than the bloom I think.

Here are some screenshots of the new version.
Native:
![v2 native](https://github.com/user-attachments/assets/f2b8c018-e05c-4477-a00c-7dead93088ad)

Recommended settings (I liked the most): bloom 50, tint 80, darken 20, desaturate 50:
![v2 recommended](https://github.com/user-attachments/assets/7c6e9b3f-4349-482c-889a-a32f99a5b8dc)

And bloom disabled with the color grading removed also (compare to disabled above):
![v2 all disabled](https://github.com/user-attachments/assets/4a8baabf-a6e0-4e6e-b56a-463c22d4a270)


I also found some HUD elements missing in the RemoveHUD pack (textbox elements) and added those for screenshots.

The only thing I would like to add still is a setting for the periodic blur of the bloom, but I didn't manage to understand those shaders yet. I guess this could be solved better by ASM, to change the SIN/time var that controls it. If anyone knows that stuff, please help!
I found some get nauseated by the blur and I feel that's the only reason they disable bloom entirely. If you could change this, the bloom could be set higher.

I still think this game is better _with_ bloom. It gives a lot of atmosphere to many places. But I agree that it is overdone a bit and more control over it feels nice.

Crementif reviewed (changes requested):
I think both the description and some of the preset names could be a little more concise and cleaned up so that general users better understand the options and the effects of this graphic pack. Other then that, your screenshots seem like they'd help a lot of people!

KoB-Kirito commented on Enhancements/TwilightPrincessHD_Bloom/rules.txt:
```diff
@@ -0,0 +1,372 @@
+titleIds = 000500001019C800,000500001019E600,000500001019E500
+name = Bloom (Glow) + Shadow World Color Grading
+path = The Legend of Zelda: Twilight Princess HD/Graphics/Bloom
+description = Bloom is used aggressively in the shadow world, where it has a strong yellow cast.||The bloom color tint is provided by the map and gives some areas their atmosphere. I would not recommend disabling bloom or its tint completely.||The shadow world has a special color grading to make the world appear darker and more colorless, which you can set independently of the bloom.||Use OpenGL  ...
```
Bloom is used everywhere in the game, not just the shadow world. There it's the most prominent though. Not sure if the description should reflect that.

Crementif replied:
Oh yeah, definitely missing that. Feel free to write a new one or modify it to include that part.

KoB-Kirito commented on Enhancements/TwilightPrincessHD_Bloom/rules.txt:
```diff
@@ -0,0 +1,372 @@
+[Definition]
+titleIds = 000500001019C800,000500001019E600,000500001019E500
+name = Bloom (Glow) + Shadow World Color Grading
+path = The Legend of Zelda: Twilight Princess HD/Graphics/Bloom
```
Bloom is everywhere, but the alpha for dampening colors and the grading shader are only used the shadow world, so the grading part is kinda only able to affect shadow world.

Crementif replied:
In my opinion its better to provide a broader graphic pack title, but if you disagree then I guess its fine.

KoB-Kirito replied:
The game uses the bloom shader for the brightness part, so you can't really seperate it.

KoB-Kirito commented on Enhancements/TwilightPrincessHD_Bloom/rules.txt:
```diff
@@ -0,0 +1,372 @@
+# Area Specific Bloom Tint (49865bd2e62efda1)
+
+[Preset]
+category = Area Specific Bloom Tint
```
Bloom is everywhere. In the shadow world the bloom tint is yellow, in the normal world at night it's blue-ish, in the mines it's orange, etc.

KoB-Kirito commented on Enhancements/TwilightPrincessHD_Bloom/rules.txt:
```diff
@@ -0,0 +1,372 @@
+# this is separated, because it is not affected by bloom strength at all
+
+[Preset]
+category = Shadow World Color Dampening
```
Hard to explain without seeing it 😅
This shader dulls the colors (only in the shadow world) by shifting everything towards the red channel, effectivly desaturating the colors. This effect at 100% desaturates like default, at 0% it doesn't desaturate at all. I wanted to give more control over the effect with more options, because it can have a dramatic effect.
Also when you are on OpenGL you can see the result live. This itself took me some time to figure out, but with that you can just try and see what it does, no?

Crementif replied:
Right, but I think it makes more sense to describe how heavy it is using textual options instead of percentage options. Maybe you can think of more text options and find some good presets? Its pretty likely that 60% of them are unusable.

From my experience, most people aren't too keen on using OpenGL (bad performance on AMD and Intel GPUs), so I'm more of a fan of having some steering and not providing too many options since it generally is overwhelming for people that are setting up a game for the first time.

KoB-Kirito replied:
You should fix live shader update for Vulkan then :P

For the name, maybe Saturation, as this option changes the Color Saturation in shadow world and the other changes the Brightness in the shadow world. Though, actually it _is_ the other way around: The game desaturates and debrightenes the color already, the setting just reverts that. I tried to reflect that like that, but you're right, maybe it's easier to understand reversed.
It would be nice if you could try the pack live and see if it still doesn't make sense then.

I disagree on the text options. Changed a lot of rules myself to get what I want (which not everyone might be capable to do), why limit it to some arbitrary presets. This is a pure setting, controlling only one aspect. I would love to just have slider for it instead of those dropdown anyways. Instead I would rather give more explaination in the description for the settings.

I get that if you don't want to tinker, a preset can help, that's why I added the recommendation comments to the ones I ended up using, as an easy starting point.

Maybe I can combine simpler presets with more custom control with conditions. I have an idea for that, but changing that on mobile isn't that convinient currently. Can try that tomorrow.

KoB-Kirito commented on Enhancements/TwilightPrincessHD_Bloom/rules.txt:
```diff
@@ -0,0 +1,372 @@
+# Shadow World Desaturation (5f422bf63e25be7f)
+
+[Preset]
+category = Shadow World Color Desaturation
```
I thought about combining both presets, but they are kinda vastly different. I myself wasn't sure for some time if I want to change both or just one, and which. So I thought more control equals better.

This one effects the actual brightness of the image. It uses the alpha value of the tint color, but only in the shadow world this ever is below 1.0.

In the end I turned this one down, to make the shadow world brighter to see better and make it less depressing, but kept the desaturation almost native, cause that help to make the shadow world more destinct.

Crementif replied:
I more so meant that I'd like to provide users with easy, high quality options rather then having them figure it out on their own. I think having multiple recommended options is a great idea. I'm also not that big of a fan of recommending options, since monitors vary wildly and what looks good on your screen might look shit on other people their screens.

KoB-Kirito commented:
So, I added simple presets and nested the finer options under a custom preset. That actually works very well.
By default you only see this and can select a preset:
![grafik](https://github.com/user-attachments/assets/13408c0e-6cc6-4df4-9168-ec5b89e0880c)
![grafik](https://github.com/user-attachments/assets/77f7f80c-5d87-4326-a184-dbce3cdafcdb)

But when you select custom, you can adjust the individual settings:
![grafik](https://github.com/user-attachments/assets/557b545f-ed63-4d8a-bae2-d673b551eb27)

Same for color grading. I also reversed the logic and the name should now make very clear what they do:
![grafik](https://github.com/user-attachments/assets/1a4b9726-6d89-4dea-9d1f-33b153613674)

Also added more to the description.
What do you think?

KoB-Kirito commented:
Here some saves to test shadow world. It takes an hour to reach it..
2. is first time shadow world (meeting midna), 3. is zanto gets midna cutscene
[save.zip](https://github.com/user-attachments/files/17383257/save.zip)

Crementif reviewed (approved):
Oh, that looks great! Thanks for making the adjustments, looks good to merge. I do have two comments, but they don't have to be addressed.
